pkg/cmd/unpack/filter.go

```diff
@@ -0,0 +1,150 @@
+	conf.filters = filters
+	return conf, nil
+}
+
+// shouldUnpackLayer determines if we should unpack a layer in a Kitfile by matching
+// fields against the filters. Matching is done against path and name (if present).
+// If filters is empty, we assume everything should be unpacked
+func shouldUnpackLayer(layer any, filters []filterConf) bool {
+	if len(filters) == 0 {
+		return true
+	}
+	// The type switch below checks for concrete (non-pointer) types. We need to use
+	// reflect to dereference the pointer and get a new interface{} (any) type.
+	if val := reflect.ValueOf(layer); val.Kind() == reflect.Ptr {
+		layer = val.Elem().Interface()
+	}
+
+	switch l := layer.(type) {
+	case artifact.KitFile:
+		for _, filter := range filters {
+			for _, baseType := range filter.baseTypes {
+				if baseType == constants.ConfigType {
+					return true
+				}
+			}
+		}
+		return false
+	case artifact.Model:
+		return matchesFilters(l.Name, constants.ModelType, filters) || matchesFilters(l.Path, constants.ModelType, filters)
+	case artifact.ModelPart:
+		return matchesFilters(l.Name, constants.ModelPartType, filters) || matchesFilters(l.Path, constants.ModelPartType, filters)
+	case artifact.Docs:
+		// Docs does not have an ID/name field so we can only match on path
+		return matchesFilters(l.Path, constants.DocsType, filters)
+	case artifact.DataSet:
+		return matchesFilters(l.Name, constants.DatasetType, filters) || matchesFilters(l.Path, constants.DatasetType, filters)
+	case artifact.Code:
+		// Code does not have a ID/name field so we can only match on path
+		return matchesFilters(l.Path, constants.CodeType, filters)
+	default:
+		return false
+	}
+}
+
+func matchesFilters(field string, baseType string, filterConfs []filterConf) bool {
+	// Treat modelparts as covered by the 'model' filter
+	if baseType == constants.ModelPartType {
+		baseType = constants.ModelType
+	}
+	for _, filterConf := range filterConfs {
+		if filterConf.matches(baseType, field) {
+			return true
+		}
+	}
+	return false
+}
+
+// filtersFromUnpackConf converts a (deprecated) unpackConf to a set of filters to enable supporting the old flags
+func filtersFromUnpackConf(conf unpackConf) []filterConf {
+	filter := filterConf{}
+
+	if conf.unpackKitfile {
+		filter.baseTypes = append(filter.baseTypes, constants.ConfigType)
+	}
+	if conf.unpackModels {
+		filter.baseTypes = append(filter.baseTypes, constants.ModelType)
+	}
+	if conf.unpackDocs {
+		filter.baseTypes = append(filter.baseTypes, constants.DocsType)
+	}
+	if conf.unpackDatasets {
+		filter.baseTypes = append(filter.baseTypes, constants.DatasetType)
+	}
+	if conf.unpackCode {
+		filter.baseTypes = append(filter.baseTypes, constants.CodeType)
+	}
+	return []filterConf{filter}
+}
+
+func filterToMediaBaseType(filterType string) (string, error) {
+	switch filterType {
+	case "kitfile":
+		return constants.ConfigType, nil
+	case "datasets":
+		// annoyingly, the mediatype is dataset, but for the filter we want the plural
```

Comment:
I think I prefer `datasets` -- while it doesn't match the mediatype, it does match the corresponding field in the Kitfile, which is going to be more familiar to users.

As an aside, I considered going for `dataset` here to make matching to a mediatype easier, but it reads kind of weird to me. The other filters, `code` and `docs`, work because they're weird words (mass/collective nouns? I'm bad at grammar but we have to consistently call them "code layers" or something else)